File: backend/models/rag_debug.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from backend.models.knowledge_access import KnowledgeAccessPolicy
from backend.rag.config import RagConfig

RagDebugRunStatus = Literal["queued", "running", "completed", "failed", "cancelled"]
RagDebugStageStatus = Literal[
    "pending", "active", "complete", "warning", "failed", "skipped"
]
RagDebugStageKey = Literal[
    "query", "rewrite", "dense", "bm25", "fusion", "rerank", "context", "answer"
]


class RagDebugModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class RagDebugCompanionTrace(RagDebugModel):
    trace_id: str
    request_id: int = Field(default=0, ge=0)
    conversation_id: str = ""
    query: str = ""
    knowledge_enabled: bool = False
    document_scope: str = "off"
    route: str
    route_reason: str = ""
    grounding_policy: str
    retrieval_skipped: bool
    verification_skipped: bool
    catalog_document_count: int = Field(default=0, ge=0)
    retrieval: dict[str, Any] = Field(default_factory=dict)
    evidence: list[dict[str, Any]] = Field(default_factory=list)
    citations: list[dict[str, Any]] = Field(default_factory=list)
    verification: dict[str, Any] = Field(default_factory=dict)
    fallback_applied: bool = False
    created_at: str


class RagDebugRunRequest(RagDebugModel):
    query: str = Field(min_length=1, max_length=4_000)
    config_id: str = Field(default="default", min_length=1, max_length=128)
    document_ids: list[str] = Field(default_factory=list, max_length=100)
    top_k: int = Field(default=8, ge=1, le=100)
    include_answer: bool = False
    workspace_id: str = Field(default="", max_length=128)
    knowledge_access_policy: KnowledgeAccessPolicy = KnowledgeAccessPolicy.AUTO


class RagDebugStageEvent(RagDebugModel):
    sequence: int = Field(ge=0)
    stage: str = Field(min_length=1, max_length=64)
    status: str = Field(min_length=1, max_length=32)
    elapsed_ms: float = Field(default=0.0, ge=0.0)
    payload: dict[str, Any] = Field(default_factory=dict)


class RagDebugStage(RagDebugModel):
    key: RagDebugStageKey
    label: str
    status: RagDebugStageStatus
    elapsed_ms: float = Field(default=0.0, ge=0.0)
    note: str = ""
    summary: dict[str, Any] = Field(default_factory=dict)
    candidate_count: int = Field(default=0, ge=0)


class RagDebugCandidate(RagDebugModel):
    id: str = Field(min_length=1)
    document_id: str = ""
    source: str = ""
    section: str = ""
    page: int | None = Field(default=None, ge=1)
    tokens: int = Field(default=0, ge=0)
    dense: float | None = None
    bm25: float | None = None
    fusion: float | None = None
    rerank: float | None = None
    before: int | None = Field(default=None, ge=1)
    after: int | None = Field(default=None, ge=1)
    text: str = ""
    chunk_type: str = ""
    start: int = Field(default=0, ge=0)
    end: int = Field(default=0, ge=0)
    metadata: dict[str, Any] = Field(default_factory=dict)


class RagDebugContext(RagDebugModel):
    text: str = ""
    estimated_tokens: int = Field(default=0, ge=0)
    included_evidence_ids: list[str] = Field(default_factory=list)
    omitted_evidence_ids: list[str] = Field(default_factory=list)
    source_count: int = Field(default=0, ge=0)


class RagDebugTraceResponse(RagDebugModel):
    run_id: str
    trace_id: str
    status: RagDebugRunStatus
    query: str
    config_id: str
    query_plan: dict[str, Any] = Field(default_factory=dict)
    stages: list[RagDebugStage] = Field(default_factory=list)
    candidates: list[RagDebugCandidate] = Field(default_factory=list)
    context: RagDebugContext = Field(default_factory=RagDebugContext)
    evidence: list[dict[str, Any]] = Field(default_factory=list)
    citations: list[dict[str, Any]] = Field(default_factory=list)
    answer: str = ""
    knowledge_decision: dict[str, Any] = Field(default_factory=dict)
    knowledge_scope: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)
    error: str = ""


class RagDebugRunAccepted(RagDebugModel):
    run_id: str
    trace_id: str
    status: RagDebugRunStatus


class RagDebugRunEventsResponse(RagDebugModel):
    run_id: str
    trace_id: str
    status: RagDebugRunStatus
    events: list[RagDebugStageEvent] = Field(default_factory=list)


class RagDebugConfigProfile(RagDebugModel):
    config_id: str
    name: str
    description: str = ""
    config: RagConfig
    active: bool = False
    requires_reindex: bool = False
    index_fingerprint: str = ""
    created_at: str
    updated_at: str


class RagDebugConfigCreate(RagDebugModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = Field(default="", max_length=500)
    config: RagConfig = Field(default_factory=RagConfig)
    activate: bool = False


class RagDebugConfigUpdate(RagDebugModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    config: RagConfig | None = None
    activate: bool | None = None


class RagDebugConfigActivation(RagDebugModel):
    config_id: str
    active: bool


class RagDebugDocument(RagDebugModel):
    document_id: str
    title: str = ""
    source_uri: str = ""
    status: str = ""
    chunk_count: int = Field(default=0, ge=0)
    updated_at: str = ""


class RagDebugChunk(RagDebugModel):
    id: str
    document_id: str
    title: str = ""
    preview: str = ""
    section: str = ""
    section_path: list[str] = Field(default_factory=list)
    page: int | None = Field(default=None, ge=1)
    tokens: int = Field(default=0, ge=0)
    overlap: int = Field(default=0, ge=0)
    start: int = Field(default=0, ge=0)
    end: int = Field(default=0, ge=0)
    type: str = ""
    embedding: str = ""
    text: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)


class RagDebugChunkPage(RagDebugModel):
    chunks: list[RagDebugChunk] = Field(default_factory=list)
    total: int = Field(ge=0)
    page: int = Field(ge=1)
    page_size: int = Field(ge=1)


class RagDebugDataset(RagDebugModel):
    dataset_id: str
    name: str
    description: str = ""
    case_count: int = Field(default=0, ge=0)
    created_at: str
    updated_at: str


class RagDebugDatasetCreate(RagDebugModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = Field(default="", max_length=500)


class RagDebugCase(RagDebugModel):
    case_id: str = Field(min_length=1, max_length=128)
    query: str = Field(min_length=1, max_length=4_000)
    categories: list[str] = Field(default_factory=list, max_length=20)
    relevant_chunk_ids: list[str] = Field(default_factory=list, max_length=100)
    relevance_grades: dict[str, int] = Field(default_factory=dict)
    claims: list[dict[str, Any]] = Field(default_factory=list, max_length=100)
    no_answer: bool = False
    # Optional routing annotations.  Older datasets remain valid: the
    # evaluator derives expected_retrieval from no_answer and derives a
    # document scope from metadata.document_id when these are omitted.
    expected_retrieval: bool | None = None
    expected_scope_document_ids: list[str] = Field(default_factory=list, max_length=100)
    metadata: dict[str, str] = Field(default_factory=dict)
    query_type: str = "Factual"
    expected_answer: str = ""
    answerable: bool = True
    tags: list[str] = Field(default_factory=list, max_length=30)
    notes: str = Field(default="", max_length=20_000)
    updated_at: str = ""


class RagDebugDatasetImport(RagDebugModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = Field(default="", max_length=500)
    content: str = Field(min_length=1, max_length=5_000_000)
    format: Literal["json", "jsonl"] | None = None


class RagDebugDatasetExport(RagDebugModel):
    dataset: RagDebugDataset
    format: Literal["json", "jsonl"]
    content: str


class RagDebugEvaluationRequest(RagDebugModel):
    dataset_id: str
    config_id: str = "default"
    top_k: int = Field(default=20, ge=1, le=100)
    case_ids: list[str] = Field(default_factory=list, max_length=500)


class RagDebugEvaluationResponse(RagDebugModel):
    dataset_id: str
    config_id: str
    report: dict[str, Any]


class RagDebugCompareRequest(RagDebugModel):
    dataset_id: str
    baseline_config_id: str = "default"
    candidate_config_id: str
    top_k: int = Field(default=20, ge=1, le=100)
    case_ids: list[str] = Field(default_factory=list, max_length=500)


class RagDebugCompareCase(RagDebugModel):
    case_id: str
    query: str
    baseline_rank: int | None = None
    candidate_rank: int | None = None
    baseline_latency_ms: float = Field(default=0.0, ge=0.0)
    candidate_latency_ms: float = Field(default=0.0, ge=0.0)
    baseline_chunk_ids: list[str] = Field(default_factory=list)
    candidate_chunk_ids: list[str] = Field(default_factory=list)


class RagDebugCompareResponse(RagDebugModel):
    dataset_id: str
    baseline_config_id: str
    candidate_config_id: str
    cases: list[RagDebugCompareCase] = Field(default_factory=list)
    metrics: dict[str, Any] = Field(default_factory=dict)


__all__ = [
    "RagDebugCandidate",
    "RagDebugCase",
    "RagDebugChunk",
    "RagDebugChunkPage",
    "RagDebugCompanionTrace",
    "RagDebugCompareCase",
    "RagDebugCompareRequest",
    "RagDebugCompareResponse",
    "RagDebugConfigActivation",
    "RagDebugConfigCreate",
    "RagDebugConfigProfile",
    "RagDebugConfigUpdate",
    "RagDebugContext",
    "RagDebugDataset",
    "RagDebugDatasetCreate",
    "RagDebugDatasetExport",
    "RagDebugDatasetImport",
    "RagDebugDocument",
    "RagDebugEvaluationRequest",
    "RagDebugEvaluationResponse",
    "RagDebugRunAccepted",
    "RagDebugRunEventsResponse",
    "RagDebugRunRequest",
    "RagDebugRunStatus",
    "RagDebugStage",
    "RagDebugStageEvent",
    "RagDebugTraceResponse",
]

File: apps/desktop/src/api/rag-debug.ts
import { apiDelete, apiGet, apiPatch, apiPost, apiWebSocketUrl } from "./client"

export type RagDebugRunStatus = "queued" | "running" | "completed" | "failed" | "cancelled"
export type RagDebugStageStatus = "pending" | "active" | "complete" | "warning" | "failed" | "skipped"
export type RagDebugTab = "trace" | "retrieval" | "chunks" | "evaluation" | "compare" | "datasets"

export interface RagConfig {
  enabled: boolean
  advanced_parsing: Record<string, unknown>
  visual_understanding: Record<string, unknown>
  visual_retrieval: Record<string, unknown>
  chunking: Record<string, unknown>
  semantic_chunking: Record<string, unknown>
  embedding: Record<string, unknown>
  vector_store: Record<string, unknown>
  retrieval: {
    dense_top_k: number
    sparse_top_k: number
    fusion_top_k: number
    final_top_k: number
    fusion: string
    small_to_big_enabled: boolean
    small_to_big_top_k: number
    small_to_big_neighbor_radius: number
    small_to_big_max_tokens_per_anchor: number
  }
  reranker: Record<string, unknown>
}

export interface RagDebugCompanionTrace {
  trace_id: string
  request_id: number
  conversation_id: string
  query: string
  knowledge_enabled: boolean
  document_scope: string
  route: string
  route_reason: string
  grounding_policy: string
  retrieval_skipped: boolean
  verification_skipped: boolean
  catalog_document_count: number
  retrieval: Record<string, unknown>
  evidence: Array<Record<string, unknown>>
  citations: Array<Record<string, unknown>>
  verification: Record<string, unknown>
  fallback_applied: boolean
  created_at: string
}

export interface RagDebugConfigProfile {
  config_id: string
  name: string
  description: string
  config: RagConfig
  active: boolean
  requires_reindex: boolean
  index_fingerprint: string
  created_at: string
  updated_at: string
}

export interface RagDebugStage {
  key: string
  label: string
  status: RagDebugStageStatus
  elapsed_ms: number
  note: string
  summary: Record<string, unknown>
  candidate_count: number
}

export interface RagDebugCandidate {
  id: string
  document_id: string
  source: string
  section: string
  page: number | null
  tokens: number
  dense: number | null
  bm25: number | null
  fusion: number | null
  rerank: number | null
  before: number | null
  after: number | null
  text: string
  chunk_type: string
  start: number
  end: number
  metadata: Record<string, unknown>
}

export interface RagDebugTraceResponse {
  run_id: string
  trace_id: string
  status: RagDebugRunStatus
  query: string
  config_id: string
  query_plan: Record<string, unknown>
  stages: RagDebugStage[]
  candidates: RagDebugCandidate[]
  context: {
    text: string
    estimated_tokens: number
    included_evidence_ids: string[]
    omitted_evidence_ids: string[]
    source_count: number
  }
  evidence: Array<Record<string, unknown>>
  citations: Array<Record<string, unknown>>
  answer: string
  knowledge_decision: Record<string, unknown>
  knowledge_scope: Record<string, unknown>
  metadata: Record<string, unknown>
  error: string
}

export interface RagDebugStageEvent {
  sequence: number
  stage: string
  status: string
  elapsed_ms: number
  payload: Record<string, unknown>
}

export interface RagDebugDocument {
  document_id: string
  title: string
  source_uri: string
  status: string
  chunk_count: number
  updated_at: string
}

export interface RagDebugChunk {
  id: string
  document_id: string
  title: string
  preview: string
  section: string
  section_path: string[]
  page: number | null
  tokens: number
  overlap: number
  start: number
  end: number
  type: string
  embedding: string
  text: string
  metadata: Record<string, unknown>
}

export interface RagDebugChunkPage {
  chunks: RagDebugChunk[]
  total: number
  page: number
  page_size: number
}

export interface RagDebugDataset {
  dataset_id: string
  name: string
  description: string
  case_count: number
  created_at: string
  updated_at: string
}

export interface RagDebugCase {
  case_id: string
  query: string
  categories: string[]
  relevant_chunk_ids: string[]
  relevance_grades: Record<string, number>
  claims: Array<Record<string, unknown>>
  no_answer: boolean
  metadata: Record<string, string>
  query_type: string
  expected_answer: string
  answerable: boolean
  tags: string[]
  notes: string
  updated_at: string
}

export interface RagDebugEvaluationResponse {
  dataset_id: string
  config_id: string
  report: {
    total_cases: number
    retrieval: Record<string, number>
    reranker: Record<string, number>
    citations: Record<string, number>
    performance: Record<string, unknown>
    cases: Array<Record<string, unknown>>
    case_details?: Array<Record<string, unknown>>
    [key: string]: unknown
  }
}

export interface RagDebugCompareCase {
  case_id: string
  query: string
  baseline_rank: number | null
  candidate_rank: number | null
  baseline_latency_ms: number
  candidate_latency_ms: number
  baseline_chunk_ids: string[]
  candidate_chunk_ids: string[]
}

export interface RagDebugCompareResponse {
  dataset_id: string
  baseline_config_id: string
  candidate_config_id: string
  cases: RagDebugCompareCase[]
  metrics: Record<string, number>
}

export interface RagDebugRunAccepted {
  run_id: string
  trace_id: string
  status: RagDebugRunStatus
}

const ROOT = "/api/rag/debug"

export function listRagDebugCompanionTraces(limit = 20): Promise<RagDebugCompanionTrace[]> {
  return apiGet(`${ROOT}/companion-traces?limit=${Math.max(1, Math.min(limit, 100))}`)
}

export function listRagDebugConfigs(): Promise<RagDebugConfigProfile[]> {
  return apiGet(`${ROOT}/configs`)
}

export function createRagDebugConfig(payload: { name: string; description?: string; config: RagConfig; activate?: boolean }): Promise<RagDebugConfigProfile> {
  return apiPost(`${ROOT}/configs`, payload)
}

export function updateRagDebugConfig(configId: string, payload: Partial<{ name: string; description: string; config: RagConfig; activate: boolean }>): Promise<RagDebugConfigProfile> {
  return apiPatch(`${ROOT}/configs/${encodeURIComponent(configId)}`, payload)
}

export function activateRagDebugConfig(configId: string): Promise<RagDebugConfigProfile> {
  return apiPost(`${ROOT}/configs/${encodeURIComponent(configId)}/activate`, {})
}

export function clearRagDebugConfigReindex(configId: string): Promise<RagDebugConfigProfile> {
  return apiPost(`${ROOT}/configs/${encodeURIComponent(configId)}/clear-reindex`, {})
}

export function deleteRagDebugConfig(configId: string): Promise<{ config_id: string; deleted: boolean }> {
  return apiDelete(`${ROOT}/configs/${encodeURIComponent(configId)}`)
}

export function listRagDebugDocuments(): Promise<RagDebugDocument[]> {
  return apiGet(`${ROOT}/documents`)
}

export function listRagDebugChunks(params: { documentId?: string; query?: string; page?: number; pageSize?: number } = {}): Promise<RagDebugChunkPage> {
  const search = new URLSearchParams()
  if (params.documentId) search.set("document_id", params.documentId)
  if (params.query) search.set("query", params.query)
  if (params.page) search.set("page", String(params.page))
  if (params.pageSize) search.set("page_size", String(params.pageSize))
  const query = search.size ? `?${search.toString()}` : ""
  return apiGet(`${ROOT}/chunks${query}`)
}

export function getRagDebugChunk(chunkId: string): Promise<RagDebugChunk> {
  return apiGet(`${ROOT}/chunks/${encodeURIComponent(chunkId)}`)
}

export function startRagDebugRun(payload: { query: string; config_id: string; document_ids?: string[]; top_k: number; include_answer: boolean; knowledge_access_policy?: "auto" | "always" | "never" }): Promise<RagDebugRunAccepted> {
  return apiPost(`${ROOT}/runs`, payload)
}

export function getRagDebugRun(runId: string): Promise<RagDebugTraceResponse> {
  return apiGet(`${ROOT}/runs/${encodeURIComponent(runId)}`)
}

export function getRagDebugRunEvents(runId: string, after = -1): Promise<{ run_id: string; trace_id: string; status: RagDebugRunStatus; events: RagDebugStageEvent[] }> {
  return apiGet(`${ROOT}/runs/${encodeURIComponent(runId)}/events?after=${after}`)
}

export function cancelRagDebugRun(runId: string): Promise<RagDebugTraceResponse> {
  return apiPost(`${ROOT}/runs/${encodeURIComponent(runId)}/cancel`, {})
}

export function ragDebugRunWebSocketUrl(runId: string): string {
  return apiWebSocketUrl(`${ROOT}/runs/${encodeURIComponent(runId)}/stream`)
}

export function listRagDebugDatasets(): Promise<RagDebugDataset[]> {
  return apiGet(`${ROOT}/datasets`)
}

export function createRagDebugDataset(payload: { name: string; description?: string }): Promise<RagDebugDataset> {
  return apiPost(`${ROOT}/datasets`, payload)
}

export function deleteRagDebugDataset(datasetId: string): Promise<{ dataset_id: string; deleted: boolean }> {
  return apiDelete(`${ROOT}/datasets/${encodeURIComponent(datasetId)}`)
}

export function listRagDebugCases(datasetId: string): Promise<RagDebugCase[]> {
  return apiGet(`${ROOT}/datasets/${encodeURIComponent(datasetId)}/cases`)
}

export function saveRagDebugCase(datasetId: string, payload: RagDebugCase): Promise<RagDebugCase> {
  return apiPost(`${ROOT}/datasets/${encodeURIComponent(datasetId)}/cases`, payload)
}

export function updateRagDebugCase(datasetId: string, caseId: string, payload: RagDebugCase): Promise<RagDebugCase> {
  return apiPatch(`${ROOT}/datasets/${encodeURIComponent(datasetId)}/cases/${encodeURIComponent(caseId)}`, payload)
}

export function deleteRagDebugCase(datasetId: string, caseId: string): Promise<{ dataset_id: string; case_id: string; deleted: boolean }> {
  return apiDelete(`${ROOT}/datasets/${encodeURIComponent(datasetId)}/cases/${encodeURIComponent(caseId)}`)
}

export function importRagDebugDataset(payload: { name: string; description?: string; content: string; format?: "json" | "jsonl" }): Promise<RagDebugDataset> {
  return apiPost(`${ROOT}/datasets/import`, payload)
}

export function exportRagDebugDataset(datasetId: string, format: "json" | "jsonl" = "json"): Promise<{ dataset: RagDebugDataset; format: "json" | "jsonl"; content: string }> {
  return apiGet(`${ROOT}/datasets/${encodeURIComponent(datasetId)}/export?format=${format}`)
}

export function evaluateRagDebugDataset(payload: { dataset_id: string; config_id: string; top_k: number; case_ids?: string[] }): Promise<RagDebugEvaluationResponse> {
  return apiPost(`${ROOT}/evaluation`, payload)
}

export function compareRagDebugDataset(payload: { dataset_id: string; baseline_config_id: string; candidate_config_id: string; top_k: number; case_ids?: string[] }): Promise<RagDebugCompareResponse> {
  return apiPost(`${ROOT}/compare`, payload)
}
